portfolio/src/main/java/com/google/sps/servlets/ImageUploaderServlet.java

```diff
@@ -0,0 +1,166 @@
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formElementName);
+
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    // We could check the validity of the file here, e.g. to make sure it's an image file
+    // https://stackoverflow.com/q/10779564/873165
+    
+    return blobKey;
+  }
+
+  /**
+   * @return the request parameter, or the default value if the parameter
+   *         was not specified by the client.
+   * Copied from TextProcessorServlet.java and modified slightly for my needs. 
+   */
+  private String getParameter(HttpServletRequest request, String parameterName, String defaultValue) {
+    String result = request.getParameter(parameterName);
+    if (result == null || result.equals("")) {
+      return defaultValue;
+    }
+    return result;
+  }
+
+  private void addToDatastore(String title, BlobKey blobKey) {
+    // Filter out bad URLs
+    if (blobKey == null) {
+      return;
+    }
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    
+    Entity imageEntity = new Entity("Image");
+    imageEntity.setProperty("title", title);
+    imageEntity.setProperty("blobKey", blobKey);
+    imageEntity.setProperty("timestamp", System.currentTimeMillis());
+
+    datastore.put(imageEntity);
+  }
+
+  private String getImagesAsJson(){
+    Gson gson = new Gson();
+
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+    Query q = new Query("Image").addSort("timestamp", SortDirection.ASCENDING);
+    PreparedQuery results = ds.prepare(q);
+
+    JSONObject images = new JSONObject(); // Contains array of "images"
+    JSONArray imgs = new JSONArray();
+
+    for (Entity e : results.asIterable()) {
+      String eAsJson = gson.toJson(e);
+      try {
+        JSONObject o = new JSONObject(eAsJson);
+        imgs.put(o);
+      } catch (JSONException jsone) {
+        // eAsJson will always be a json because it's generated by gson
+        // so we never actually reach this line
+      }
+    }
+
+    try {
+      images.put("images", imgs);
+    } catch (JSONException jsone) {
+      // imgs will always be a json because it's the string form of a JSONArray. 
+      // It's impossible to get to this line. 
+    }
+
+    return images.toString();
```

Comment:
Can you please clean the code up a little bit by use JsonObject and JsonArray from Gson to replace ones from JSON.
```suggestion 
  private String getImagesAsJson(){
    Gson gson = new Gson();
    JsonArray imgs = new JsonArray();

    for (Entity e : results.asIterable()) {
      imgs.put(gson.toJson(e));
    }

    JsonObject images = new JsonObject();
    images.add("images", imgs);
    return images.getAsString();
  }
```